miden-lib/asm/sat/internal/account_vault.masm

```diff
@@ -63,3 +66,227 @@ export.has_non_fungible_asset
     movdn.4 dropw movdn.4 dropw
     # => [has_asset]
 end
+
+# ADD ASSET
+# =================================================================================================
+
+# TODO: Replace with an implementation that uses the advice provider to give new amount for
+#       insertion. Ideally we shouldn't be putting vault logic in miden-vm so how can we do
+#       this ergonomically?
+
+#! Add the specified fungible asset to the vault.  If the vault already contains an asset
+#! issued by the same faucet, the amounts are added together.
+#!
+#! Panics:
+#! - If the total value of assets is greater than or equal to 2^63.
+#!
+#! Stack: [ASSET]
+#! Output: [ASSET']
+#!
+#! - ASSET is the fungible asset to add to the vault.
+#! - ASSET' is the updated fungible asset.
+proc.add_fungible_asset
+    push.0 movdn.3
+    # => [ASSET_KEY, amount]
+
+    # get the asset vault root and read asset
+    exec.layout::get_acct_vault_root swapw exec.smt::get
+    # => [CUR_ASSET, VAULT_ROOT, amount]
+
+    # arrange elements
+    movup.3 movup.8 dup
+    # => [amount, amount, cur_amount, faucet_id, 0, 0, VAULT_ROOT]
+
+    # compute max_amount - cur_amount
+    exec.asset::get_fungible_asset_max_amount dup.3 sub
+    # => [(max_amount - cur_amount), amount, amount, cur_amount, faucet_id, 0, 0, VAULT_ROOT]
+
+    # asset amount + cur_amount < max_amount
+    lte assert
+    # => [amount, cur_amount, faucet_id, 0, 0, VAULT_ROOT]
```

Comment:
Technically, this check is not needed as we should ensure no fungible asset can have supply of over $2^{63}$ (this would be enforced by the `mint` procedure). But it also may be a good redundant check (in case we are not taking something into account). So, may we should keep it for now and create an issue to remove this in the future?

Reply:
Great point.  I think your proposal makes sense.  Lets leave it for now and then remove it once we have more confidence in the protocol.